src/zql/ast-to-ivm/pipeline-builder.test.ts

```diff
@@ -95,5 +97,162 @@ test('Where', () => {
+        ],
+      },
+      expected: ['a', 'c', 'd'],
+    },
+    {
+      name: 'basic (1) OR condition',
+      where: {
+        op: 'OR',
+        conditions: [{op: '=', field: 'a', value: {type: 'literal', value: 1}}],
+      },
+      expected: ['a', 'c'],
+    },
+    {
+      name: 'basic (3) OR condition',
+      where: {
+        op: 'OR',
+        conditions: [
+          {op: '=', field: 'a', value: {type: 'literal', value: 1}},
+          {op: '=', field: 'b', value: {type: 'literal', value: 2}},
+          {op: '=', field: 'a', value: {type: 'literal', value: 2}},
+        ],
+      },
+      values: [
+        {id: 'a', a: 1, b: 1},
+        {id: 'b', a: 2, b: 1},
+        {id: 'c', a: 1, b: 2},
+        {id: 'd', a: 3, b: 3},
+      ],
+      expected: ['a', 'c', 'b'],
+    },
+    {
+      name: 'two branches with same condition',
+      where: {
+        op: 'OR',
+        conditions: [
+          {op: '=', field: 'a', value: {type: 'literal', value: 1}},
+          {op: '=', field: 'a', value: {type: 'literal', value: 1}},
+        ],
+      },
+      expected: ['a', 'c'],
+    },
+    {
+      name: 'WHERE (a = 1 AND b = 1) OR (a = 2 AND b = 2)',
+      where: {
+        op: 'OR',
+        conditions: [
+          {
+            op: 'AND',
+            conditions: [
+              {op: '=', field: 'a', value: {type: 'literal', value: 1}},
+              {op: '=', field: 'b', value: {type: 'literal', value: 1}},
+            ],
+          },
+          {
+            op: 'AND',
+            conditions: [
+              {op: '=', field: 'a', value: {type: 'literal', value: 2}},
+              {op: '=', field: 'b', value: {type: 'literal', value: 2}},
+            ],
+          },
+        ],
+      },
+      expected: ['a', 'd'],
+    },
+    {
+      name: 'WHERE (a = 1 OR b = 1) AND (a = 2 OR b = 2)',
+      where: {
+        op: 'AND',
+        conditions: [
+          {
+            op: 'OR',
+            conditions: [
+              {op: '=', field: 'a', value: {type: 'literal', value: 1}},
+              {op: '=', field: 'b', value: {type: 'literal', value: 1}},
+            ],
+          },
+          {
+            op: 'OR',
+            conditions: [
+              {op: '=', field: 'a', value: {type: 'literal', value: 2}},
+              {op: '=', field: 'b', value: {type: 'literal', value: 2}},
+            ],
+          },
+        ],
+      },
+      expected: ['b', 'c'],
+    },
+  ];
+
+  for (const c of cases) {
+    test(c.name, () => {
+      const {values = defaultValues} = c;
+      const m = new Materialite();
+      const s = m.newStatelessSource();
+      const ast: AST = {
+        table: 'e1',
+        select: ['id'],
+        where: c.where,
+        orderBy: [['id'], 'asc'],
+      };
+
+      const pipeline = buildPipeline(() => s.stream, ast);
+
+      // console.log(pipeline.toString());
+
+      const log: unknown[] = [];
+
+      // why in the world is effect not notified but debug is?!?!?!?!?!
```

Comment:
so the problem is that `effect` only notifies on commit (intentional) but it doesn't retain values it has seen in the current transaction.

`or` -> `concat` -> `distinct` emits twice:
- Once for the left branch feeding into it
- Once for the right branch

If the row passes both `OR` conditions then:
1. The first time it emits a row,
2. The second time it does not emit a row

Effect throws out its old value and replaces it with nothing at step (2), causing effect to notify with empty results.

tldr: effect should retain all values it sees then notify with them on commit.